More Marvin commands

thokkat opened the pull request:
This PR implements marvin keys `F4-F7` and the console command `toggle time` https://github.com/Try/OpenGothic/issues/215.

F4 - Reset camera to normal mode.
F5 - Freeze camera. Camera stays in place and can't be moved. Character  can move freely.
F6 - Free camera. Allows to fly the camera through the world. Character is frozen.
F7 - Like normal play but camera is shifted to a position reached by F6. Camera moves with the character.

F4-F6 are implemented  the same as vanilla. There `F7` aligns the camera to the players head-to-foot axis. I haven't looked into that and kept the usual y-axis, but can try if it's desirable. 

I encountered some hiccups when the camera changes to a different mode. I opted to not stop marvin cam in those cases for now, there's always `F4` to reset.

Comparison F7 PR/vanilla

https://user-images.githubusercontent.com/100468436/217689362-e630d643-764a-4393-a6a0-3b304f9e7dcd.mp4

https://user-images.githubusercontent.com/100468436/217688735-b3a97cdc-b140-4f9f-956a-4d3c3a5a3ec7.mp4

Try commented:
Hi, @thokkat , thanks for PR!

Can you please give more detail explanation on each command? Expected vanilla behavior and implemented one(if differs)

thokkat commented:
Edited start post.

Try commented on game/camera.cpp:
```diff
@@ -145,6 +150,45 @@ void Camera::setLookBack(bool lb) {
+    cameraOffset.y = offset.y;
+    }
+  else
+    dst.target = origin;
```
nitpick: inconsistent brackets. Should be
```
if(...) {
  foo();
  } else {
  boo();
  }
```

Try reviewed (commented):
I've made a quick only look so far, will do testing, once I'm done with npc swimming code, that I'm currently working on.

Try commented on game/camera.h:
```diff
@@ -31,6 +31,13 @@ class Camera final {
+      M_Normal,
+      M_Freeze,
+      M_Free,
+      M_Fixed,
```
M_Fixed
> Like normal play but camera is shifted to a position reached by F6. Camera moves with the character.

Name is misleading. Should it be, for example, "pinned" ?

Try commented on game/camera.h:
```diff
@@ -117,6 +127,7 @@ class Camera final {
     bool                  lbEnable      = false;
     bool                  inertiaTarget = true;
     Mode                  camMod        = Normal;
+    MarvinMode            camMarvinMod  = M_Normal;
```
Are `Mode` and  `MarvinMode` complementary or mutably exclusive?
If exclusive - it make sense to have them into one enum.

thokkat replied:
This is a bit unclear to me. If a dialog is started in F5/F7 mode dialog cam takes over and afterwards F5/F7 takes over again. So there has to be some memory which cam was active before.
 
I moved the marvinMode enum to `mainwindow.cpp` and use `solveCameraMode` routine  to return the respective camera. So only one enum in `camera.cpp` with marvin cams included.

Try replied:
> f a dialog is started in F5/F7 mode dialog cam takes over and afterwards F5/F7 takes over 

I didn't know that - resolved

Try commented on game/world/worldobjects.cpp:
```diff
@@ -161,9 +161,15 @@ void WorldObjects::tick(uint64_t dt, uint64_t dtPlayer) {
-      npc.tick(dtPlayer); else
+    if(npc.isPlayer()) {
+      if(Gothic::inst().camera() && Gothic::inst().camera()->isMarvinMode(Camera::M_Free)) {
+        return;
```
Also why player is not receiving ticks, when in free-cam mode?

thokkat replied:
Player can't move, doesn't drop if in air, receives no damage if diving bar is empty, `aigoto` only starts if `free cam` mode is left and current animation is frozen. Player can be damaged by monsters and npc though but death animation behaves like `aigoto`. Maybe it's best to just activate `godmode` here. 

So basically all what's done in player-tick is skipped/ignored.

Try replied:
Tested vanilla today:
In free cam move playable npc doesn't receive animation ticks for skeletal animation, an no physics ticks. It can receive damage and perceptions(magic damage at least).
Also main menu not working in free-camera O-o

Anyway `return;` is definitely wrong here.

Try commented on game/mainwindow.cpp:
```diff
@@ -687,16 +704,32 @@ void MainWindow::onMarvinKey() {
+          camera->setMarvinMode(Camera::M_Freeze);
+          }
+        } else {
+        Gothic::inst().quickSave();
```
While this is correct, not having `F5` in debug play would be painfully annoying. Maybe remove `camera-freeze`, from debug build?

Try commented on game/camera.cpp:
```diff
@@ -122,6 +125,32 @@ void Camera::setMode(Camera::Mode m) {
+  if(m==M_Free)
+    pl->setAnim(AnimationSolver::NoAnim);
+  if(m==M_Fixed) {
+    auto offset    = dst.target - pl->cameraBone();
```
This math odd. 
Is it another way to write `inverse(pl->cameraBone()) * dst.target` ?

thokkat replied:
This is supposed to calculate the vector from player to camera position. Is there a better way?

thokkat replied:
Isn't it what I'm currently doing?  Code below would lead to same offset vector.
```
   Vec3 offset(0,0,1);
   Matrix4x4 m;
   m.identity();
   m.translate(pl->cameraBone(true));
   m.inverse();
   m.translate(dst.target);
   m.project(offset);  
   cameraOffset = offset;
```

Try replied:
I mean it make sense to change it, and fetch `cameraBone` from npc as matrix, with tilt and rotation.

thokkat replied:
Edit: Looks like `offsetAng` could do the tilt.

Try commented on game/camera.cpp:
```diff
@@ -122,6 +125,32 @@ void Camera::setMode(Camera::Mode m) {
+    cameraOffset.z = offset.x * s - offset.z * c;
+    cameraOffset.y = offset.y;
+    }
+  else if(camMarvinMod==M_Fixed) {
```
This line is misleading.
2 previous condition in a row were about `m`(new mode), suggesting switch-case like workflow, but here it's suddenly `camMarvinMod`.

ezamelczyk commented:
I can't really pinpoint why it's happening but when I build this branch on m1 mac I have an issue. Menus are working but in the game everything's funky. I preformed a clean build and it did not help. Main branch does not have this issue. Screenshot attached.

![Zrzut ekranu 2023-02-28 o 11 54 07](https://user-images.githubusercontent.com/9358409/221833607-7d9ad81d-d8c4-4cfd-8aa5-8d6e4b64fd2f.png)

thokkat commented:
@ezamelczyk 
Everything this PR does is disabled by default. Once activated only the camera position should change. Maybe it's because this pr was some weeks behind master. But master is now merged in so you could give it another try.

@Try
 I addressed the requested changes. I don't know if there's a better way to prevent animation ticks. It seems the animation is reset  instead of continued if e.g `F6->F7`, which leads to problems with mobsi states. As a workaround i prevented the use of F6 in these cases.

Also didn't manage to get tilt in. Everything i did also changed the rotation axis. I'd like to leave that for a follow pr, except you know right away how to do it. Although i like it how it is, just a change of camera perspective.

ezamelczyk commented:
Yup. Seems to be working correctly with updates from main.

Try commented:
Checkout with github no longer working :(
Anyway, code look ok - will merge it as-is now, and fix nitpicks in post
